contracts/FragToken.sol

```diff
@@ -115,27 +109,15 @@ contract FRAGToken is ERC20, ERC20Permit, Ownable{
     function unlock(bytes calldata signature) external {
 
         uint256 amount = 0;
-        uint256 len = _userlocks[msg.sender].length();
-        uint256[] memory toRemove = new uint256[](len);
-
         // loop over all the locks performed by the sender and calculate the aggregate unlockable
-        for (uint i = 0; i < len; i++) {
-            uint256 struct_hash = _userlocks[msg.sender].at(i); 
-
-            if(_lockInfos[struct_hash].locktime < block.timestamp) {
-                amount += _lockInfos[struct_hash].amount;
-
-                toRemove[i] = struct_hash;
+        for (uint i = _lockInfos[msg.sender].length - 1; i >= 0 ; i--) {
```

Comment:
`_lockInfos[msg.sender].length` this one you might wanna use a temporary var, as it might do hashmap seek.